modules/navier_stokes/src/fvkernels/CNSFVHLLCBase.C

```diff
@@ -0,0 +1,119 @@
+  params.addRequiredParam<UserObjectName>(nms::fluid, "Fluid userobject");
+  return params;
+}
+
+CNSFVHLLCBase::CNSFVHLLCBase(const InputParameters & params)
+  : FVFluxKernel(params),
+    _fluid(dynamic_cast<FEProblemBase *>(&_subproblem)
+               ->getUserObject<SinglePhaseFluidProperties>(nms::fluid)),
+    _specific_internal_energy_elem(getADMaterialProperty<Real>(nms::specific_internal_energy)),
+    _specific_internal_energy_neighbor(
+        getNeighborADMaterialProperty<Real>(nms::specific_internal_energy)),
+    _rho_et_elem(getADMaterialProperty<Real>(nms::total_energy_density)),
+    _rho_et_neighbor(getNeighborADMaterialProperty<Real>(nms::total_energy_density)),
+    _vel_elem(getADMaterialProperty<RealVectorValue>(nms::velocity)),
+    _vel_neighbor(getNeighborADMaterialProperty<RealVectorValue>(nms::velocity)),
+    _speed_elem(getADMaterialProperty<Real>(nms::speed)),
+    _speed_neighbor(getNeighborADMaterialProperty<Real>(nms::speed)),
+    _rho_elem(getADMaterialProperty<Real>(nms::density)),
+    _rho_neighbor(getNeighborADMaterialProperty<Real>(nms::density)),
+    _pressure_elem(getADMaterialProperty<Real>(nms::pressure)),
+    _pressure_neighbor(getNeighborADMaterialProperty<Real>(nms::pressure))
+{
+  if (_tid == 0)
+    _fi_to_wave_speeds.resize(libMesh::n_threads());
+}
+
+HLLCData
+CNSFVHLLCBase::hllcData() const
+{
+  return {_fluid,
+          _rho_elem[_qp],
+          _rho_neighbor[_qp],
+          _vel_elem[_qp],
+          _vel_neighbor[_qp],
+          _specific_internal_energy_elem[_qp],
+          _specific_internal_energy_neighbor[_qp]};
+}
+
+const std::array<ADReal, 3> &
+CNSFVHLLCBase::waveSpeed(const THREAD_ID tid,
+                         const FaceInfo & fi,
+                         const HLLCData & hllc_data,
+                         const ADRealVectorValue & normal)
+{
+  mooseAssert(tid < libMesh::n_threads(), tid << " must be less than " << libMesh::n_threads());
+
+  auto it = _fi_to_wave_speeds[tid].find(&fi);
+  if (it != _fi_to_wave_speeds[tid].end())
+    return it->second;
+
+  const ADReal & rho1 = hllc_data.rho_elem;
+  const ADReal u1 = hllc_data.vel_elem.norm();
+  const ADReal q1 = normal * hllc_data.vel_elem;
+  const ADReal v1 = 1.0 / rho1;
+  const ADReal & e1 = hllc_data.e_elem;
+  const ADReal E1 = e1 + 0.5 * u1 * u1;
+  const ADReal p1 = hllc_data.fluid.p_from_v_e(v1, e1);
+  const ADReal H1 = E1 + p1 / rho1;
+  const ADReal c1 = hllc_data.fluid.c_from_v_e(v1, e1);
+
+  const ADReal & rho2 = hllc_data.rho_neighbor;
+  const ADReal u2 = hllc_data.vel_neighbor.norm();
+  const ADReal q2 = normal * hllc_data.vel_neighbor;
+  const ADReal v2 = 1.0 / rho2;
+  const ADReal & e2 = hllc_data.e_neighbor;
+  const ADReal E2 = e2 + 0.5 * u2 * u2;
+  const ADReal p2 = hllc_data.fluid.p_from_v_e(v2, e2);
+  const ADReal H2 = E2 + p2 / rho2;
+  const ADReal c2 = hllc_data.fluid.c_from_v_e(v2, e2);
+
+  // compute Roe-averaged variables
```

Comment:
well I m up for reading chapter 11 I guess

Reply:
I also believe I documented all of this in `CNSFVHLLCBase.md` ...

Reply:
(hopefully). A second pair of eyes is probably good

Reply:
Ok chapter 11 wasnt the relevant one. It s in chapter 10. 
So there's tons of options here for determining SL and SR.

I dont see this one in the list (in my edition). It s essentially a mix between Davis' "exceedingly simple" ones and the recommended by Davis and Einfeldt Roe average. 

Toro seems to recommend a pressured based wave speed estimate. But his equations are written for ideal gases, and you have to choose between two different approximations, more choices..

So I m happy with this, and if we notice issues with wave speeds, we ll have to re-visit

Reply:
> I dont see this one in the list (in my edition). It s essentially a mix between Davis' "exceedingly simple" ones and the recommended by Davis and Einfeldt Roe average.

Yea...I noted the same thing when I took up this code. @snschune presumably wrote this so maybe he got it from another reference? It is slightly unsatisfactory to have a wave speed implementation without a corresponding reference.

Reply:
so you were saying you had to go with a smaller time step that you expected right. Maybe this is the problem? Like the "exceedingly simple" option is being more limiting for the faster wave speed

Because the averaging done by Roe is always 'in the middle' since it s an average, like q_roe is smaller than q_2 and c_roe is probably the same as c_2 roughly.

If you have time, could you please try to confirm my intuition that we just want to use Roe and not this. 
This is also just a wild intuition so you dont have to spend the time.

@snschune if you provide a reference on this, we dont need to do this

Reply:
Note that CFL is a necessary but not necessarily sufficient condition for stability

Reply:
My hypothesis would be that the stability limitation is not due to the wave speed estimates, but it would be good to give it a whirl and verify. I will do this if you don't beat me to it @GiudGiud

Reply:
I just checked on the 1D shock tube. Pure Roe vs Roe blend shows little difference. The max time step I can take with pure Roe is 0.086 while with Roe blend it's 0.087. I ll do a little more testing but now that we have a reference for our wave speeds that's good with me.

Reply:
It's a bit funny that the Einfeldt paper was published before HLLC was even created and we use it as a reference for signal speeds. (Toro did HLLC in 1994). *But* Batten published in 1997 and he clearly discusses HLLC so I think that's a "solid" reference